ibis/expr/operations.py

```diff
@@ -2525,272 +2337,329 @@ class ExtractMillisecond(ExtractTimestampField):
-
-
-class TimeAdd(Add):
-
-    input_type = [rules.time, rules.interval(units=['h', 'm', 's'])]
-    output_type = rules.shape_like_arg(0, 'time')
 
+class TimeAdd(BinaryOp):
+    left = Arg(rlz.time)
+    right = Arg(rlz.interval(units={'h', 'm', 's', 'ms', 'us', 'ns'}))
+    output_type = rlz.shape_like('left')
 
-class TimeSub(Subtract):
 
-    input_type = [rules.time, rules.interval(units=['h', 'm', 's'])]
-    output_type = rules.shape_like_arg(0, 'time')
+class TimeSub(BinaryOp):
+    left = Arg(rlz.time)
+    right = Arg(rlz.interval(units={'h', 'm', 's', 'ms', 'us', 'ns'}))
+    output_type = rlz.shape_like('left')
 
 
 class TimeDiff(BinaryOp):
+    left = Arg(rlz.time)
+    right = Arg(rlz.time)
+    output_type = rlz.shape_like('left', dt.Interval('s'))
 
-    input_type = [rules.time, rules.time]
-    output_type = rules.shape_like_arg(0, dt.Interval('s'))
 
+class TimestampAdd(BinaryOp):
+    left = Arg(rlz.timestamp)
+    right = Arg(rlz.interval(units={'Y', 'Q', 'M', 'W', 'D',
+                                    'h', 'm', 's', 'ms', 'us', 'ns'}))
+    output_type = rlz.shape_like('left')
 
-class TimestampAdd(Add):
 
-    input_type = [rules.timestamp, rules.interval]
-    output_type = rules.shape_like_arg(0, 'timestamp')
-
-
-class TimestampSub(Subtract):
-
-    input_type = [rules.timestamp, rules.interval]
-    output_type = rules.shape_like_arg(0, 'timestamp')
+class TimestampSub(BinaryOp):
+    left = Arg(rlz.timestamp)
+    right = Arg(rlz.interval(units={'Y', 'Q', 'M', 'W', 'D',
+                                    'h', 'm', 's', 'ms', 'us', 'ns'}))
+    output_type = rlz.shape_like('left')
 
 
 class TimestampDiff(BinaryOp):
-
-    input_type = [rules.timestamp, rules.timestamp]
-    output_type = rules.shape_like_arg(0, dt.Interval('s'))
+    left = Arg(rlz.timestamp)
+    right = Arg(rlz.timestamp)
+    output_type = rlz.shape_like('left', dt.Interval('s'))
 
 
-class IntervalAdd(Add):
-
-    input_type = [rules.interval, rules.interval]
+class IntervalAdd(BinaryOp):
+    left = Arg(rlz.interval)
+    right = Arg(rlz.interval)
 
     def output_type(self):
-        left, right = self.args
-
-        # should upconvert to the smaller unit?
-        # value_type = left.to_unit(right.type().unit).type()
-        return rules.shape_like(left, left.type())
+        args = [arg.cast(arg.type().value_type) for arg in self.args]
+        expr = rlz.numeric_like(args, operator.add)(self)
+        dtype = dt.Interval(self.left.type().unit, expr.type())
+        return rlz.shape_like(self.args, dtype=dtype)
 
 
-class IntervalMultiply(Multiply):
+class IntervalMultiply(BinaryOp):
+    left = Arg(rlz.interval)
+    right = Arg(rlz.numeric)
 
     def output_type(self):
-        helper = rules.IntervalPromoter(self.left, self.right, operator.mul)
-        return helper.get_result()
-
-
-class IntervalFloorDivide(FloorDivide):
+        args = [self.left.cast(self.left.type().value_type), self.right]
```

Comment:
A couple local variables wouldn't kill you here :)

Reply:
Well, it indeed needs a cleanup.